go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,220 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"reflect"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+// getACLName returns the ACL name if it has one otherwise returns
+// an empty string.
+func getACLName(acl *nbdb.ACL) string {
+	if acl.Name != nil {
+		return *acl.Name
+	}
+	return ""
+}
+
+// isEquivalentACL if it hase same uuid, or if it has same name
+// and external ids, or if it has same priority, direction, match
+// and action.
+func isEquivalentACL(existing *nbdb.ACL, searched *nbdb.ACL) bool {
+	if searched.UUID != "" && existing.UUID == searched.UUID {
+		return true
+	}
+
+	eName := getACLName(existing)
+	sName := getACLName(searched)
+	// TODO if we want to support adding/removing external ids,
+	// we need to compare them differently, perhaps just the common subset
+	if eName != "" && eName == sName && reflect.DeepEqual(existing.ExternalIDs, searched.ExternalIDs) {
+		return true
+	}
+
+	return existing.Priority == searched.Priority &&
+		existing.Direction == searched.Direction &&
+		existing.Match == searched.Match &&
+		existing.Action == searched.Action
+}
+
+// findACL looks up the client cache for equivalent ACLs
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) (string, error) {
+	if acl.UUID != "" && !isNamedUUID(acl.UUID) {
+		return acl.UUID, nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return "", fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return "", fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 1 {
+		return acls[0].UUID, nil
+	}
+
+	return "", nil
+}
+
+func BuildACL(name, direction string, priority int, match, action, meter, severity string, log bool, externalIds map[string]string) *nbdb.ACL {
```

Comment:
If this is going to be higher-order function, then I'd like to see different types for all the strings. This sort of string soup is always a pain.

Model-gen has already taken care of most of this for you. There are types ACLAction, ACLDirection, and ACLSeverity. When we bump to the 21.06 schema, it will include constants for the acceptable values as well.

Reply:
> Separately, when is this useful? Why not just fold this in to CreateOrUpdateACL, which takes an ACL literal

Right now it is useful to have a place to operate commonly over ACL data, like ACL name crop or named UUIDs.

> and supplies the UUID as necessary?

I made a conscious decision to avoid this. Tried to keep uuid handling within libovsdbops given that uuids are not always available, the need to handle named uuids and the fact that we might need other ACL data to do lookups when uuid is not available. So I guessed that basing the libovsdbops api homogenously on the structs (that might contain a valid real uuid or not) rather than uuids or selective data would be more flexible. uuids on and by themselves are not really useful, only for deletes I guess. Also you won't get back the uuid immediately if you plan on batching.

So far this worked well for ACLS and PGs because we didn't need any additional caching but we might need to some extra stuff for other types. 

Like can having a `TransactAndCheckAndSetUUIDs` that given the models that intervened in the operation updates them with the real uuid based on the named uuids used in the operations provided and the obtained results. And that should work regardless of batching or no batching and while not free still cheaper than making expensive libovsdb cache lookups and perhaps cheaper than being forced to a no-batching option.